pkg/nfd-topology-gc/nfd-nrt-gc.go

```diff
@@ -0,0 +1,219 @@
+	}
+
+	clientset := kubernetes.NewForConfigOrDie(config)
+	factory := informers.NewSharedInformerFactory(clientset, 5*time.Minute)
+
+	return &nrtGarbageCollector{
+		topoClient: cli,
+		stopChan:   stop,
+		gcPeriod:   gcPeriod,
+		factory:    factory,
+	}, nil
+}
+
+func (n *nrtGarbageCollector) deleteNodeHandler(object interface{}) {
+	// handle a case when we are starting up and need to clear stale NRT resources
+	obj := object
+	if deletedFinalStateUnknown, ok := object.(cache.DeletedFinalStateUnknown); ok {
+		klog.V(2).Infof("found stale NodeResourceTopology for node: %s ", object)
+		obj = deletedFinalStateUnknown.Obj
+	}
+
+	node, ok := obj.(*corev1.Node)
+	if !ok {
+		klog.Errorf("cannot convert %v to v1.Node", object)
+		return
+	}
+	if err := n.topoClient.TopologyV1alpha1().NodeResourceTopologies().Delete(context.TODO(), node.GetName(), metav1.DeleteOptions{}); err != nil {
+		if errors.IsNotFound(err) {
+			klog.V(2).Infof("NodeResourceTopology for node %s not found, omitting deletion", node.GetName())
+			return
+		} else {
+			klog.Warningf("failed to delete NodeResourceTopology for node %s: %s", node.GetName(), err.Error())
+			return
+		}
+	}
+	klog.Infof("NodeResourceTopology for node %s has been deleted", node.GetName())
+}
+
+// populateNodeIndexer populates cache with NRTs so we know which one to delete on first synchronization
+func (n *nrtGarbageCollector) populateNodeIndexer(indexer cache.Indexer) error {
+	nrts, err := n.topoClient.TopologyV1alpha1().NodeResourceTopologies().List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+
+	for _, nrt := range nrts.Items {
+		node := &corev1.Node{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: nrt.GetName(),
+			},
+		}
+		klog.V(2).Infof("adding node %s to indexer", node.GetName())
+		err := indexer.Add(node)
+		if err != nil {
+			return err
+		}
+
+	}
+
+	return nil
+}
+
+// periodicGC runs garbage collector at every gcPeriod to make sure we haven't missed any node
+func (n *nrtGarbageCollector) periodicGC(gcPeriod time.Duration) {
+	indexer := n.factory.Core().V1().Nodes().Informer().GetIndexer()
+	gcTrigger := time.NewTicker(gcPeriod)
+	for {
+		select {
+		case <-gcTrigger.C:
+			klog.Infof("Running periodic GC")
+			objects := indexer.List()
+			nodes := map[string]struct{}{}
+			for _, object := range objects {
+				key, err := cache.MetaNamespaceKeyFunc(object)
+				if err != nil {
+					klog.Warningf("cannot create key for %v: %s", object, err.Error())
+					continue
+				}
+				nodes[key] = struct{}{}
+			}
+			nrts, err := n.topoClient.TopologyV1alpha1().NodeResourceTopologies().List(context.TODO(), metav1.ListOptions{})
+			if err != nil {
+				klog.Warningf("cannot list NRTs %s", err.Error())
+				break
+			}
+			for _, nrt := range nrts.Items {
+				key, err := cache.MetaNamespaceKeyFunc(&nrt)
+				if err != nil {
+					klog.Warningf("cannot create key for %v: %s", nrt, err.Error())
+					continue
+				}
+				if _, ok := nodes[key]; !ok {
```

Comment:
(not-so-minor but not blocking either) in this case I'd try to extract a helper to delete the NRT given a nodeName, to be reused both here and in `deleteNodeHandler`

Reply:
done.